test/test_dynamic_type.cpp

```diff
@@ -666,4 +770,113 @@ TEST_ASSIGN_OP(>>, >>=, RShiftAssign);
 #pragma clang diagnostic pop
 #endif
 
+namespace container_test {
+// Testing containers support by implementing the set-theoretic definition of
+// natural numbers:
+// https://en.wikipedia.org/wiki/Set-theoretic_definition_of_natural_numbers
+
+// TODO: unordered set is a better fit for this case, but it does not work with
+// some old compilers (for example the old gcc on our CI). This is a workaround
+#if !defined(__GNUC__) || __GNUC__ >= 12
+
+struct StupidHash {
+  template <typename T>
+  size_t operator()(const T&) const {
+    // This hash always collides, but who cares?
+    return 0;
+  }
+};
+
+template <typename T>
+using UnorderedSetWithStupidHash = std::unordered_set<T, StupidHash>;
+
+#else
+
+template <typename T>
+using UnorderedSetWithStupidHash = std::vector<T>;
+#define insert push_back
+
+#endif
+
+using NaturalNumber = DynamicType<Containers<UnorderedSetWithStupidHash>>;
+
+using Set = UnorderedSetWithStupidHash<NaturalNumber>;
+
+#if 1
+// DynamicType doesn't support operator== for containers, so we need to define
+// it ourselves.
+// TODO: add support for containers operator overloading in DynamicType, and
+// remove this definition.
+
+// operator== has to be in the top namespace for clang, otherwise clang will
+// have trouble compiling it. operator== has to be in the container_test
+// namespace for gcc, otherwise gcc will not compile.
```

Comment:
This is super ugly, but in my followup PRs, operator overloading for containers will be supported, so this entire `#if 1` block will be removed. So I don't really care it is ugly right now.

Reply:
Fixed in https://github.com/NVIDIA/Fuser/pull/523